scripts/task.ts

```diff
@@ -31,166 +58,404 @@ export type TemplateDetails = {
       await outputFile(junitFilename, prefixedXml);
     }
   }
 }
 
 async function run() {
-  const {
-    task: taskKey,
-    template: templateKey,
-    force,
-    before,
-    junit,
-  } = await getOptionsOrPrompt('yarn task', options);
-
-  return runTask(taskKey, templateKey, {
-    mustBeReady: false,
-    mustNotBeReady: force,
-    before,
-    junit,
-  });
+  const allOptionValues = await getOptionsOrPrompt('yarn task', options);
+
+  const { task: taskKey, startFrom, junit, ...optionValues } = allOptionValues;
+
+  const finalTask = tasks[taskKey];
+  const { template: templateKey } = optionValues;
+  const template = TEMPLATES[templateKey];
+  const templateSandboxDir = templateKey && join(sandboxDir, templateKey.replace('/', '-'));
+  const details = {
+    key: templateKey,
+    template,
+    codeDir,
+    sandboxDir: templateSandboxDir,
+    builtSandboxDir: templateKey && join(templateSandboxDir, 'storybook-static'),
+    junitFilename: junit && getJunitFilename(taskKey),
+  };
+
+  const { sortedTasks, tasksThatDepend } = getTaskList(finalTask, optionValues);
+  const sortedTasksReady = await Promise.all(
+    sortedTasks.map((t) => t.ready(details, optionValues))
+  );
+
+  logger.info(`Task readiness up to ${taskKey}`);
+  const initialTaskStatus = (task: Task, ready: boolean) => {
+    if (task.service) {
+      return ready ? 'serving' : 'notserving';
+    }
+    return ready ? 'ready' : 'unready';
+  };
+  const statuses = new Map<Task, TaskStatus>(
+    sortedTasks.map((task, index) => [task, initialTaskStatus(task, sortedTasksReady[index])])
+  );
+  writeTaskList(statuses);
+
+  function setUnready(task: Task) {
+    // If the task is a service we don't need to set it unready but we still need to do so for
+    // it's dependencies
+    if (!task.service) statuses.set(task, 'unready');
+    tasksThatDepend
+      .get(task)
+      .filter((t) => !t.service)
+      .forEach(setUnready);
+  }
+
+  // NOTE: we don't include services in the first unready task. We only need to rewind back to a
+  // service if the user explicitly asks. It's expected that a service is no longer running.
+  const firstUnready = sortedTasks.find((task) => statuses.get(task) === 'unready');
+  if (startFrom === 'auto') {
+    // Don't reset anything!
+  } else if (startFrom === 'never') {
+    if (!firstUnready) throw new Error(`Task ${taskKey} is ready`);
+    if (firstUnready !== finalTask)
+      throw new Error(`Task ${getTaskKey(firstUnready)} was not ready`);
+  } else if (startFrom) {
+    // set to reset back to a specific task
+    if (firstUnready && sortedTasks.indexOf(tasks[startFrom]) > sortedTasks.indexOf(firstUnready)) {
+      throw new Error(
+        `Task ${getTaskKey(firstUnready)} was not ready, earlier than your request ${startFrom}.`
+      );
+    }
+    setUnready(tasks[startFrom]);
+  } else if (firstUnready === sortedTasks[0]) {
```

Comment:
can this be moved to the top of this if/else chain?

Reply:
I wanted to make sure if you specific `startFrom=never` or `startFrom=X` where X is after the first task it throws an error in this case.